Add securityContext config for Redis to helm chart

dan-vaughan opened the pull request:
- Add securityContext templating to statefulSet manifest
- Add securityContext commented-out to values.yaml
- Add securityContext section to values.schema.json

Commit:
Add securityContext config for Redis to helm chart

- Add securityContext templating to statefulSet manifest
- Add securityContext commented-out to values.yaml
- Add securityContext section to values.schema.json

jedcunningham reviewed (changes requested):
Thanks @flipstone42!

Can you add the redis statefulset to the test suite for this feature?

https://github.com/apache/airflow/blob/2499da8f1e477c94e1a52bdb60b8151c09024363/tests/charts/test_security_context.py#L41

dan-vaughan committed:
Merge branch 'apache:main' into helm-add-redis-securitycontext

jedcunningham committed:
Test coverage

jedcunningham commented on chart/templates/redis/redis-statefulset.yaml:
```diff
@@ -22,6 +22,7 @@
 {{- $nodeSelector := or .Values.redis.nodeSelector .Values.nodeSelector }}
 {{- $affinity := or .Values.redis.affinity .Values.affinity }}
 {{- $tolerations := or .Values.redis.tolerations .Values.tolerations }}
+{{- $securityContext := include "airflowSecurityContext" (list . .Values.redis) }}
```
```suggestion
{{- $securityContext := include "localSecurityContext" .Values.redis }}
```

This should use `localSecurityContext` instead. We don't want the global Airflow security context to apply to Redis.

jedcunningham commented on chart/values.schema.json:
```diff
@@ -3764,6 +3764,19 @@
                             }
                         }
                     }
+                },
```
```suggestion
                },
                "uid": {
                    "description": "Redis run as user parameter.",
                    "type": "integer",
                    "default": 999
                },
```

Which means we also need `uid`.

jedcunningham commented on chart/values.schema.json:
```diff
@@ -3764,6 +3764,19 @@
+                    "default": {},
+                    "examples": [
+                        {
+                            "runAsUser": 50000,
```
```suggestion
                            "runAsUser": 999,
```

This will be a better example, as this is the uid of the `redis` user.

jedcunningham commented on chart/values.yaml:
```diff
@@ -1328,6 +1328,11 @@ redis:
   affinity: {}
   tolerations: []
 
+  # When not set, the values defined in the global securityContext will be used
```
```suggestion
  uid: 999
  # If not set, `redis.uid` will be used
```

jedcunningham replied:
I'll add that using `999` as the default does change the behavior as it runs as `redis` not `root`, but that worked for me even if the redis db was owned as `root`. If we want to be extra safe though, we might consider defaulting to `0` instead.

dan-vaughan replied:
Probably safer to default to 0 and make a note that this is for backwards-compatibility

jedcunningham commented on tests/charts/test_security_context.py:
```diff
@@ -33,6 +33,7 @@ def test_check_deployments_and_jobs(self):
             },
             show_only=[
                 "templates/flower/flower-deployment.yaml",
+                "templates/redis/redis-statefulset.yaml",
```
```suggestion
```

jedcunningham replied:
Instead, we should refactor `test_check_statsd_uid` to be like this:

```
    # Test containerSecurity priority over uid under components using localSecurityContext
    def test_check_local_uid(self):
        component_contexts = {"uid": 3000, "securityContext": {"runAsUser": 7000}}
        docs = render_chart(
            values={
                "redis": {**component_contexts},
                "statsd": {"enabled": True, **component_contexts},
            },
            show_only=[
                "templates/statsd/statsd-deployment.yaml",
                "templates/redis/redis-statefulset.yaml",
            ],
        )

        for doc in docs:
            assert 7000 == jmespath.search("spec.template.spec.securityContext.runAsUser", doc)
```

jedcunningham commented on chart/values.yaml:
```diff
@@ -1328,6 +1328,11 @@ redis:
 
+  # When not set, the values defined in the global securityContext will be used
+  securityContext: {}
+  #  runAsUser: 50000
```
```suggestion
  #  runAsUser: 999
```

Commit:
Implement @jedcunningham's recommended changes

- Set uid (default 0) in values
- Use localSecurityContext in redis statefulset
- Refactor test for localSecurityContext

Commit:
Implement @jedcunningham's recommended changes

- Set uid (default 0) in values
- Use localSecurityContext in redis statefulset
- Refactor test for localSecurityContext

Commit:
Merge branch 'helm-add-redis-securitycontext' of github.com:flipstone42/airflow into helm-add-redis-securitycontext

dan-vaughan committed:
Merge branch 'main' into helm-add-redis-securitycontext

Commit:
Add note on uid

Commit:
Merge branch 'helm-add-redis-securitycontext' of github.com:flipstone42/airflow into helm-add-redis-securitycontext

dan-vaughan committed:
Merge branch 'main' into helm-add-redis-securitycontext

Commit:
add redis statefulset to test_check_local_setting

dan-vaughan committed:
Merge branch 'main' into helm-add-redis-securitycontext

Commit:
fix path to statefulset manifest

Commit:
Merge branch 'helm-add-redis-securitycontext' of github.com:flipstone42/airflow into helm-add-redis-securitycontext

dan-vaughan committed:
Merge branch 'main' into helm-add-redis-securitycontext
